### PR TITLE
Stack protobuf storage

### LINKED ISSUE #164: Stack Protobuf storage
relates to https://github.com/appcelerator/amp/issues/162
Store stacks as protocol buffers

follow on to https://github.com/appcelerator/amp/issues/161

### PR DESCRIPTION
**Pull request by generalhenry**
closes https://github.com/appcelerator/amp/issues/164

Defines services and stacks as in the stack.proto, casts parsed format into them and stores them on stack create (and up)

### CODE REVIEW

File: api/rpc/stack/parse.go [hunk 1]
```diff
@@ -9,23 +9,13 @@ import (
 type serviceMap struct {
 	Image       string      `yaml:"image"`
 	Ports       []string    `yaml:"ports"`
-	Replicas    int         `yaml:"replicas"`
+	Replicas    int32       `yaml:"replicas"`
```
Comment (subfuzion) [inline api/rpc/stack/parse.go]: I know it seems excessively large (I thought so as well), but the actual type used by docker for `replicas` is a `uint64`, and it's what I'm using for service right now.

File: api/rpc/stack/parse.go [hunk 2]
```diff
@@ -9,23 +9,13 @@ import (
...
-	Ports       map[nat.Port][]nat.PortBinding
-	Replicas    int
-	Environment map[string]string
-	Labels      map[string]string
-}
-
-func parseStackYaml(in string) (out []Service, err error) {
-	out = []Service{}
+func parseStackYaml(in string) (out Stack, err error) {
+	out = Stack{}
```
Comment (subfuzion) [inline api/rpc/stack/parse.go]: Minor thing, but I'd save the value copying for the return and do this instead;  `out = &Stack{}`

File: api/rpc/stack/stack.go [hunk 3]
```diff
@@ -7,34 +7,37 @@ import (
 	"golang.org/x/net/context"
 )
 
-// Stack is used to implement stack.StackServer
-type Stack struct {
+// Server is used to implement stack.Server
+type Server struct {
```
Comment (subfuzion) [inline api/rpc/stack/stack.go:10]: I definitely don't prefer calling it `Server`. Conflict with some other type, like a message?

**Review by subfuzion — CHANGES_REQUESTED** [review]
Looks good overall, but a few questions or things to address...

File: api/rpc/stack/parse.go (hunk 1 above)
Comment (generalhenry) [reply api/rpc/stack/parse.go]: done

File: api/rpc/stack/parse.go (hunk 2 above)
Comment (generalhenry) [reply api/rpc/stack/parse.go]: done

File: api/rpc/stack/stack.go (hunk 3 above)
Comment (subfuzion) [reply api/rpc/stack/stack.go:10]: I got the rationale, thanks.

**Review by subfuzion — APPROVED** [review]

**Comment by subfuzion** [comment]
LGTM, thanks 👍

**Pull request merged**